### ISSUE TITLE
Create JS library

### ISSUE THREAD

**Issue by gauteh** [astroidmail/astroid#403]
## Basic
* [X] Load JS library
* [x] Define message structure
* [x] Serialize message using JSON structure (#404)

## Messages
* [X] Add message
   - [ ] Add code tags in order to get syntax highlighting either if message is patch or if code tags are found
* [ ] Add attachments
   - [ ] Show thumbnails
* [ ] Shows marked state
* [ ] Toggle states
* [ ] Update messages (#332)
* [x] Insert messages
* [ ] Re-sort messages (#398)
* [ ] Toggle indentation 

## Navigation
* [ ] Navigation
* [ ] Scrolling
* [ ] Focusing

**Comment by gauteh** [comment]
@Ramblurr

**Comment by Ramblurr** [comment]
Can you describe the high level way that the C bits and JS bits will interact?

What data does the C layer need to pass to the JS layer? And how does it pass it?

What actions/data does the JS send up to the C layer? 

I need to browse through the current source too. 

We should also get a branch building with webkit 2.0, even if it doesn't render anything, do you think that's possible?

**Comment by Ramblurr** [comment]
First proof of concept is up at: https://github.com/Ramblurr/astroid/tree/webkit2js

It includes:

* basic rendering of focused email message (html by default, with images loaded)
* basic crendering of non-focused message
* email body content is sandboxed using an [iframe sandbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox)

Many things aren't rendered yet:

* marked status
* `attachment_template`
* `encrypt_template`
* `sibling_template`
* `mime_template`
* `link_warning_template`  - what is this?
* spinner isn't used

## js compilation
I've included a [README](https://github.com/Ramblurr/astroid/blob/webkit2js/ui/README.md) in the ui dir and a simple Makefile.

The Makefile should be considered temporary, if you are satisfied with the direction I'm taking, we can migrate it to python/SConstruct later.

The Makefile will, by default attempt to use the google closure compiler to compile/optimize the javascript. The google closure compiler is a java program, which is packaged in most distros as `closure-compiler`. 

Using it isn't required, I've provided a way to bundle the library without the compiler (see the README). However, using it is highly recommended as it does minification and unused code removal. Currently it is saving 292.89 KB. This is because I use a helper library (ramda) but only use a few functions of the hundreds it offers.

Generally speaking I think this setup with closure-compiler is preferred over any sort of setup that requires nodejs, npm, and all their *lovely* friends.

## data structures & js api

[Here](https://github.com/Ramblurr/astroid/blob/webkit2js/ui/astroid.js#L217) you can see the structures I've defined for the message and related structures that the C++ layer would provide. These are open to discussion of course.

[You can also see](https://github.com/Ramblurr/astroid/blob/webkit2js/ui/astroid.js#L261) the public API that the js layer will expose to the C++ layer. So far it is just `init`, `add_message` and `clear_messages`.

### stubbing

I've provided an example of the the message structure and the api in the [stub.js](https://github.com/Ramblurr/astroid/blob/webkit2js/ui/stub.js#L9) file.

## next steps

In addition to all the tasks in the top post:

* implement plain-text mode
* implement image toggling in html mode
* stub out and render the remaining pieces (see above).

**Comment by gauteh** [comment]
Nice! this looks great!! I didn't have time right now to test everything out, but I started to try using JSON to generate the message structure, but I don't think I managed to call the JS API correctly, my attempts here: https://github.com/gauteh/astroid/tree/ramjs

Looking forward to test this more! Looks very good.

**Comment by Ramblurr** [comment]
The js code isn't forgiving at the moment, all of the fields on the message are required. 

I fleshed out the C++ code and now we have a working example of adding messages: https://github.com/Ramblurr/astroid/tree/ramjs

You'll likely want to clean that up :) Also the preview and body are not included, as I'm not sure the best way to get those.

**Comment by gauteh** [comment]
Thanks! I've cleaned things up a little bit and put it in a branch on the main repo (webkit2js), I'll invite you to the organization so that you also have access to that branch and keep that as a stable base for this work.

**Comment by gauteh** [comment]
I think most of the message structure is passed to the JS now (including attachments and mime-messages), let's try to keep missing things about that specifically in #404.

A few thoughts:
* Tags need to have foreground and background color passed in from the C++ layer (this is sometimes provided by plugins so we can't just re-implement it in JS, need to extend the structure to accommodate this), the following structure should work:
```
tags - array of tag objects
tag {
tag - string
fg - color string (#XXXXXX)
bg - color string (#XXXXXX)
```
I can put this into the webkit2js branch.

Scrolling and focus:
* Can scrolling be handled by the JS library?
* When scrolling down, we sometimes change focus to make sure the focus follows the visible messages.
* And; when focusing down, we sometimes need to scroll in order to make the focused message visible

These depend on the size of the elements (and whether they are expanded or not - the .hide tag) so they require access to the DOM tree. If this cannot be handled entirely by the JS lib we need to start passing things back from JS. If that is the case, the best would be to just pass back a scroll request.

Message state:
* Expanded
* Marked
* Focus

In order to update these state parameters, perhaps the best would be to pass in a map from C++ with message-id + state for each of the parameters. In all cases the parameter will only change for one message (or element in the case of focus) at the time (except perhaps at initial render). If no element within the message is focused the message is considered to be the "root-element" which then has the focus.

**Comment by Ramblurr** [comment]
> * Expanded

What is expanded?

> In all cases the parameter will only change for one message (or element in the case of focus) at the time

What do you mean by element here?

**Comment by Ramblurr** [comment]
I just discovered what elements are! I've never used that feature in Astroid before

Using j and k one can scroll between the elements, and enter opens it in a new tab.

So what can be an element? So far I've seen:

* MIME messages
* Alternative parts
* Attachments
* Paragraphs (?)

The paragraph bit is puzzling. How does that work?

To handle this in the JS side, each element will need some sort of id / reference.

**Comment by gauteh** [comment]
The elements are defined in `ThreadView::MessageState`, but `Address` and `Encryption` were not used yet. Essentially, anything that should be navigable to by keyboard. If no element is selected, then the entire message is in focus (this is then the first "virtual" element within each message). This has relevance e.g. when you try to save the current element: should the whole message be saved or just the current attachment.

**Comment by Ramblurr** [comment]
Can we  pass the unique integer  id for the  elements to JS?

When Astroid requests that an element be focused, JS needs to know which element that  is.

**Comment by gauteh** [comment]
Done.

**Comment by gauteh** [comment]
It seems that it is not trivial to do synchronous JS callbacks, thread: https://lists.webkit.org/pipermail/webkit-gtk/2017-October/003167.html

**Comment by Ramblurr** [comment]
Quick update @gauteh, my internet has been out for awhile now :/ hopefully it will be fixed soon and I can get back to work.

**Comment by gauteh** [comment]
Casey Link writes on oktober 24, 2017 10:20:
> Quick update @gauteh, my internet has been out for awhile now :/ hopefully it will be fixed soon and I can get back to work.
> 

Ok! Good to hear! Realized I hadn't heard from you in a while. I haven't 
had the chance to familiarize myself with the JS code myself yet, so 
wasn't really able to make any real progress on that part.

**Comment by Ramblurr** [comment]
@gauteh Already, diving in again.

I have a few questions about the initialization sequence of the thread view.

There are at least XX pieces of interaction state that need to be initialized and managed:

1. the focused message
2. the focused element
3. and the expanded flag for each message

What should the initial values of these be?

Should the values be passed by C++ to JS?
Or, should the JS layer choose? If the latter.. then how should it choose?

**Comment by Ramblurr** [comment]
Ok! Can you add the C++ code / stubs that will call the js with the initial state? It would help me implement the rest of the focusing/expanding functionality.

It can either be passed in one structure, or individually, but if possible I would recommend passing all the data in one call. I've been using:

```js

{ // object keyed by message id
<mid>: { focused: boolean, expanded: boolean }
}

// example 
{
"XXXXX": { focused: false, expanded: true}
"YYYYY": { focused: false, expanded: false}
}
```

**Comment by gauteh** [comment]
Casey Link writes on oktober 26, 2017 17:24:
> Ok! Can you add the C++ code / stubs that will call the js with the initial state? It would help me implement the rest of the focusing/expanding functionality.
> 
> It can either be passed in one structure, or individually, but if possible I would recommend passing all the data in one call. I've been using:
> 
> ```js
> 
> { // object keyed by message id
> <mid>: { focused: boolean, expanded: boolean }
> }
> 
> // example 
> {
> "XXXXX": { focused: false, expanded: true}
> "YYYYY": { focused: false, expanded: false}
> }
> ```

Ok, I am now doing it per-message by calling:

* `Astroid.expand_message (mid)` and 
* `Astroid.collapse_message (mid)` 

If you wish I could pass an entire map with the expanded status each 
time. But, this will typically be manipulated on one message at the 
time.

* All messages are collapsed by default.
* No messages are focused by default.

**Comment by gauteh** [comment]
Closed in favor of #455. Thanks for the effort!
